src/kernel/kernel.h

```diff
@@ -233,10 +233,11 @@ static inline void __attribute__((always_inline)) context_acquire(context ctx, c
     context_debug("%s: ctx %p, cpu %d\n", __func__, ctx, ci->id);
     assert(ctx->active_cpu != ci->id);
     u64 remain = CONTEXT_RESUME_SPIN_LIMIT;
-    while (!compare_and_swap_32(&ctx->active_cpu, -1u, ci->id)) {
+    while (ctx->active_cpu != -1u) {
```

Comment:
This will be fine provided that there is never more than one closure scheduled to run for a given context. This should be the case as contextual closures should be used for a series of _sequential_ operations that occur under a given context. There could be cases where parallel operations are issued on behalf of a context (e.g. multiple I/O operations that fan out from a single syscall and then merge together into a completion) but those usually just require the context's fault handler to be borrowed in order to allow user pages to be faulted in (via use_fault_handler()).

It's essential that there is never more than one cpu trying to acquire a context at a time, for without an atomic acquire it is possible that a context could be scheduled on two different cpus simultaneously. This could lead to undefined behavior that might be difficult to debug. I'm not sure off hand how we could detect or assert against such a condition.